### PR TITLE
DOCR-280: rebase upstream distribution

### PR DESCRIPTION
**Pull request by dhij**
Jira: DOCR-280

This is a rebase on upstream/main. Since the rebase involves a huge number of commits, I summarized the outstanding changes on our end vs. their end (upstream) in `s3.go` below, since that is where the most overlap happens. 

Passed e2e-public. How I ran e2e (including here just to make sure it is correct):
- On the data plane, I replaced the pseudo version of the replace directive inside go mod with `replace github.com/docker/distribution => github.com/digitalocean/docker-distribution v0.0.0-20230314030308-90987cd4d3c5` which references the last commit of this PR (now different commit hash since I cleaned up commits), and ran `go mod tidy` and `go install`
- Ran `dev/e2e-public` on the control plane

#### Added on our end
`s3.go`:
- Using HeadObject and falling back to ListObject in Stat()
    - https://github.com/distribution/distribution/commit/3f2424cf0fdc398cfcf3240702a9caa68496d071 & https://github.com/distribution/distribution/commit/556d80827448e15088d79da0010d0bc1c08239d1
- Using S3 params
    - https://github.com/distribution/distribution/commit/aabf07ec657fdca1138783edf9f4700d36993af8
- LogS3APIRequests & LogS3APIResponseHeaders as driver params
    - https://github.com/digitalocean/docker-distribution/commit/f3d61f10008fc5bcda9edc0b3093a73e74009229
    - Using `d.s3Client(ctx)` instead of calling d.S3 directly
- QuotaExceeded Error
    - https://github.com/digitalocean/docker-distribution/commit/d192a974fbede4b4f868ff99bcfff2bdc235b297
- setContentLength
    - https://github.com/digitalocean/docker-distribution/commit/b0fc5ddc8b3a7e64ca4dd1830d12a3419eea14a9 & https://github.com/digitalocean/docker-distribution/commit/bdbf05f6f8e8708084023b10f0ff35dbafd90a59

Other commits:
- Add Vacuum.RemoveLayerLink() https://github.com/distribution/distribution/commit/4366eee39ade8de0acef831c300f88b770da1018
- Remove layer links along with manifest https://github.com/distribution/distribution/commit/3b7741805d84c3414f9c8b9aa891b8a565f89a09
- Add ModTime to registry blob descriptor https://github.com/distribution/distribution/commit/0e4a61bbae7ec0983759f83acc50f4b6c2959ca6

#### Added on their end (upstream) 
`s3.go`:

- Fixed max layer size of 10GB bug
    - https://github.com/distribution/distribution/pull/2815
- Paginate through s3 multipart uploads - fixed multipart uploads over 1,000 rows
    - https://github.com/distribution/distribution/pull/3519
    - Related issue: [DOCR-259](https://jira.internal.digitalocean.com/browse/DOCR-259)
- UseDualStack, MultipartCombineSmallPart, Accelerate, & ForcePathStyle as driver params
    - https://github.com/digitalocean/docker-distribution/commit/e2caaf9cba7c119868865ec35aa02cd0297ac810, https://github.com/digitalocean/docker-distribution/commit/117757a5cb8ae931f5eab4de24763d8c0421c522, https://github.com/digitalocean/docker-distribution/commit/51c0c8148ab75fae72f60957b3b9f246e421b6f5, https://github.com/digitalocean/docker-distribution/commit/15de9e21bad774b24e48a46d9238a5714e7ceb6c
- Add all supported storage classes
    - https://github.com/digitalocean/docker-distribution/commit/fb937deabf5845996d06a5d7edd7c1b5146346c8 & https://github.com/digitalocean/docker-distribution/commit/966fae54639c0268949ee9a329b6951b92b4921f
- Update to ListObjectsV2
    - https://github.com/digitalocean/docker-distribution/commit/8eab5d1bd6c69e895d762c30a0d34031ccc872ee
- Delete S3 keys incrementally in batches
    - https://github.com/digitalocean/docker-distribution/commit/ebc4234fd573e21ee49b0a7b14e5e76818a6a7b0

Other information:
- There are a lot of commits that Collin added on the `doWalk()` in `s3.go` inside the 170 commits that we are ahead by [here](https://github.com/distribution/distribution/compare/main...digitalocean:docker-distribution:master) (Jun 30, 2021). Those confused me but later realized there is a [PR](https://github.com/distribution/distribution/pull/3480) he merged into upstream so I believe those are the most up-to-date changes for `doWalk()`.

### CODE REVIEW

File: registry/storage/linkedblobstore.go [hunk 1]
```diff
@@ -46,7 +46,7 @@ type linkedBlobStore struct {
 var _ distribution.BlobStore = &linkedBlobStore{}
 
 func (lbs *linkedBlobStore) Stat(ctx context.Context, dgst digest.Digest) (distribution.Descriptor, error) {
-	return lbs.blobAccessController.Stat(ctx, dgst)
+	return lbs.blobStore.statter.Stat(ctx, dgst)
```
Comment (dhij) [inline registry/storage/linkedblobstore.go:49]: Related thread [here](https://digitalocean.slack.com/archives/CMXK6M39D/p1678127940419069) where I was asking about the unknown/missing blob.

If I understood it correctly, [Wayne’s PR](https://github.com/digitalocean/docker-distribution/pull/5) made it not possible to pull some images that could be missing layer links and caused a problem. So you reverted the change [here](https://github.com/distribution/distribution/commit/37ab304645167e4103c58e05100ece6a42796cd0). It looks like our docker-distribution still keeps the `lbs.blobStore.statter.Stat()` which references the actual blob so we should not have problem fetching a blob without the layer links.

My thoughts now are that:
1. Given how we get the 404 `blob unknown to registry` despite using `lbs.blobStore.statter.Stat()`, I am assuming  [DOCR-590](https://jira.internal.digitalocean.com/browse/DOCR-590) is a different problem.
2. As to whether it is now safe to use `lbs.blobAccessController.Stat()`, I need to look further into upstream code to see any relevant fixes.

**Review by adamwg — COMMENTED** [review]
Overall this looks fine to me, given that you've tested it in the DOCR context already. A few things that caught my attention but may be fine:

1. A couple of commits are duplicated in the rebase - e.g., Bouke's `registry: convert AWS S3 error to correct errocde` and its reversion. Any idea how that ended up happening? Probably it would be good to remove the later version.
2. Similarly, it looks like at least one of my commits (099d9e8ededf276556aa363d9b3e98701b1ad930 in this branch) ended up with no actual content, but some conflict markers committed. I suspect the actual contents of this commit were merged upstream, and the conflict markers were fixed in a fixup later in the branch. Assuming that's the case, I'd generally be in favor of dropping any commits that were merged upstream so that the history is tidier and it's clearer what patches we're carrying in our fork.
3. I recall some discussion with Dylan about the need to re-rename the module back to `github.com/docker/distribution`, but I don't recall where we landed on it. Ideally it would be nice to leave the rename to `distribution/distribution` in place so we're not carrying a giant patch that affects so many files. Do you know what breaks if we don't rename? Maybe Billie can help us figure out a better solution, given his more extensive go modules experience.

**Comment by dhij** [comment]
@adamwg Thank you for your feedback!
> 
> 1. A couple of commits are duplicated in the rebase - e.g., Bouke's `registry: convert AWS S3 error to correct errocde` and its reversion. Any idea how that ended up happening? Probably it would be good to remove the later version.

I believe the duplication is because of Dylan's previous attempt at merging/rebasing upstream e.g. this [commit](https://github.com/digitalocean/docker-distribution/commit/90aff9a48a52e1062c21eececafffb06bb871558). Although the whole [PR was reverted](https://github.com/digitalocean/docker-distribution/pull/29), those commits that Dylan had merged conflicts for exist in history. I can clean up the duplicates.

> 2. Similarly, it looks like at least one of my commits ([099d9e8](https://github.com/digitalocean/docker-distribution/commit/099d9e8ededf276556aa363d9b3e98701b1ad930) in this branch) ended up with no actual content, but some conflict markers committed. I suspect the actual contents of this commit were merged upstream, and the conflict markers were fixed in a fixup later in the branch. Assuming that's the case, I'd generally be in favor of dropping any commits that were merged upstream so that the history is tidier and it's clearer what patches we're carrying in our fork.

Thank you for pointing that out. I will make the changes accordingly 🙂

> 3. I recall some discussion with Dylan about the need to re-rename the module back to `github.com/docker/distribution`, but I don't recall where we landed on it. Ideally it would be nice to leave the rename to `distribution/distribution` in place so we're not carrying a giant patch that affects so many files. Do you know what breaks if we don't rename? Maybe Billie can help us figure out a better solution, given his more extensive go modules experience.

Oh, I'd thought that we wanted to maintain the `github.com/docker/distribution` as the module name in our forked repository since that is how our current [go.mod](https://github.com/digitalocean/docker-distribution/blob/master/go.mod) has. Do you advise that we name our module `distribution/distribution/v3` and keep the import paths as they are? 

I recall the issue Dylan was having was with /v3 suffix causing problem with the replace directive in our docr data plane. But he seemed to have resolved it with something like `replace github.com/distribution/distribution/v3 => github.com/digitalocean/docker-distribution/v3 v3.0.0-20230118122740-06562fdeeb39` as seen in PR #107 on our DOCR data plane (which I am not putting the link here since this is a public repo) though that PR wasn't merged.

**Comment by dhij** [comment]
Re-doing the rebase on a separate PR https://github.com/digitalocean/docker-distribution/pull/35

**Pull request closed without merging**